### PR TITLE
[ADF-643] upload enhancements

### PR DESCRIPTION
**Pull request by DenisVuyka**
Numerous upload enhancements:

- limits concurrent downloads not to block UI
- flattens hierarchy on folder upload
- performs a single traversal of the entire folder hierarchy and ends with a complete file list
- allows now dropping folders on existing folders
- overall code improvements
- cleanup demo shell demo for the document list

Marked for deprecation:

- UploadButtonComponent.showNotificationBar
- UploadButtonComponent.currentFolderPath
- UploadDragAreaComponent.showNotificationBar
- UploadDragAreaComponent.currentFolderPath

### CODE REVIEW

File: ng2-components/ng2-alfresco-core/src/directives/upload.directive.ts [hunk 1]
```diff
@@ -215,7 +246,11 @@ export class UploadDirective implements OnInit, OnDestroy {
         if (this.isClickMode()) {
             const input = (<HTMLInputElement>e.currentTarget);
             const files = this.getFilesSelected(input.files);
-            this.onUploadFiles(files);
+            this.onUploadFiles(files.map(f => <FileInfo> {
```
Comment (mauriziovitale) [inline ng2-components/ng2-alfresco-core/src/directives/upload.directive.ts]: can you change f with fileInfo

Comment (DenisVuyka) [inline ng2-components/ng2-alfresco-core/src/directives/upload.directive.ts]: updated the code

File: ng2-components/ng2-alfresco-upload/src/components/upload-button.component.ts [hunk 2]
```diff
@@ -180,37 +155,24 @@ export class UploadButtonComponent implements OnInit, OnChanges {
...
-     * @param path
      * @param files
+     * @param path
      */
-    uploadFiles(path: string, files: File[]): void {
-        if (files.length) {
-            const latestFilesAdded = files.map(f => new FileModel(f, { newVersion: this.versioning }));
+    uploadFiles(files: File[]): void {
+        if (files.length > 0) {
+            const latestFilesAdded = files.map(f => new FileModel(f, {
```
Comment (eromano) [inline ng2-components/ng2-alfresco-upload/src/components/upload-button.component.ts]: one letter variable

File: ng2-components/ng2-alfresco-upload/src/components/upload-drag-area.component.ts [hunk 3]
```diff
@@ -67,15 +73,18 @@ export class UploadDragAreaComponent {
...
-                if (e.detail.data.obj.entry.isFolder) {
-                    let id = e.detail.data.obj.entry.id;
-                    this.onFilesDropped(fileModels, id, '/');
-                } else {
-                    this.onFilesDropped(fileModels);
+                let parentId = this.rootFolderId;
+                if (e.detail.data && e.detail.data.obj.entry.isFolder) {
+                    parentId = e.detail.data.obj.entry.id || this.rootFolderId;
                 }
+                const fileModels = files.map(f => new FileModel(f.file, {
```
Comment (eromano) [inline ng2-components/ng2-alfresco-upload/src/components/upload-drag-area.component.ts]: one letter variable

File: demo-shell-ng2/app/components/files/files.component.ts [hunk 4]
```diff
@@ -160,24 +110,11 @@ export class FilesComponent implements OnInit, AfterViewInit {
...
-        });
-    }
-
-    private getBpmActionHandler(processDefinition: any): ContentActionHandler {
-        return function (obj: any, target?: any) {
-            window.alert(`Starting BPM process: ${processDefinition.id}`);
-        }.bind(this);
+    onFileUploadComplete(event: FileUploadCompleteEvent) {
+        if (event && event.file.options.parentId === this.documentList.currentFolderId) {
+            console.log('Reloading on File upload complete');
```
Comment (eromano) [inline demo-shell-ng2/app/components/files/files.component.ts]: do we need this console log?

File: ng2-components/ng2-alfresco-upload/README.md [hunk 5]
```diff
@@ -177,11 +123,11 @@ platformBrowserDynamic().bootstrapModule(AppModule);
 | Name | Type | Default | Description |
 | --- | --- | --- | --- |
 | `disabled` | *boolean* | false | Toggle component disabled state |
-| `showNotificationBar` | *boolean* | true | Hide/show notification bar |
+| **(deprecated)** `showNotificationBar` | *boolean* | true | Hide/show notification bar. **Deprecated in 1.6.0: use UploadService events and NotificationService api instead.** |
 | `uploadFolders` | *boolean* | false | Allow/disallow upload folders (only for chrome) |
 | `multipleFiles` | *boolean* | false | Allow/disallow multiple files |
 | `acceptedFilesType` | *string* | * |  array of allowed file extensions , example: ".jpg,.gif,.png,.svg" |
-| `currentFolderPath` | *string* | '/Sites/swsdp/documentLibrary' | define the path where the files are uploaded |
+| **(deprecated)** `currentFolderPath` | *string* | '/Sites/swsdp/documentLibrary' | define the path where the files are uploaded. **Deprecated in 1.6.0: use rootFolderId instead.** |
```
Comment (eromano) [inline ng2-components/ng2-alfresco-upload/README.md]: rootFolderIdproperties missing in the table

File: ng2-components/ng2-alfresco-upload/README.md [hunk 6]
```diff
@@ -300,9 +209,9 @@ platformBrowserDynamic().bootstrapModule(AppModule);
 | Name | Type | Default | Description |
 | --- | --- | --- | --- |
 | `enabled` | *boolean* | true | Toggle component enabled state |
-| `showNotificationBar` | *boolean* | true |  Hide/show notification bar |
+| **(deprecated)** `showNotificationBar` | *boolean* | true |  Hide/show notification bar. **Deprecated in 1.6.0: use UploadService events and NotificationService api instead.** |
 | `rootFolderId` | *string* | '-root-' | The ID of the root folder node.
```
Comment (eromano) [inline ng2-components/ng2-alfresco-upload/README.md]: rootFolderId should not be in the events

Comment (DenisVuyka) [inline ng2-components/ng2-alfresco-upload/README.md]: it's a properties table going after events

File: ng2-components/ng2-alfresco-upload/README.md (hunk 5 above)
Comment (DenisVuyka) [inline ng2-components/ng2-alfresco-upload/README.md]: fixed

File: demo-shell-ng2/app/components/files/files.component.ts (hunk 4 above)
Comment (DenisVuyka) [inline demo-shell-ng2/app/components/files/files.component.ts]: removed console.log

File: ng2-components/ng2-alfresco-upload/src/components/upload-drag-area.component.ts (hunk 3 above)
Comment (DenisVuyka) [inline ng2-components/ng2-alfresco-upload/src/components/upload-drag-area.component.ts]: fixed

File: ng2-components/ng2-alfresco-upload/src/components/upload-button.component.ts (hunk 2 above)
Comment (DenisVuyka) [inline ng2-components/ng2-alfresco-upload/src/components/upload-button.component.ts]: fixed

**Comment by DenisVuyka** [comment]
@eromano made all changes requested and fixed unit tests

**Pull request merged**